File: src/main/java/pl/swmud/ns/swaedit/gui/BalloonWidget.java
package pl.swmud.ns.swaedit.gui;

import com.trolltech.qt.core.QPoint;
import com.trolltech.qt.core.QRectF;
import com.trolltech.qt.core.QTimer;
import com.trolltech.qt.core.Qt;
import com.trolltech.qt.gui.QBitmap;
import com.trolltech.qt.gui.QBrush;
import com.trolltech.qt.gui.QCloseEvent;
import com.trolltech.qt.gui.QColor;
import com.trolltech.qt.gui.QContentsMargins;
import com.trolltech.qt.gui.QFont;
import com.trolltech.qt.gui.QImage;
import com.trolltech.qt.gui.QLabel;
import com.trolltech.qt.gui.QMouseEvent;
import com.trolltech.qt.gui.QPaintEvent;
import com.trolltech.qt.gui.QPainter;
import com.trolltech.qt.gui.QPainterPath;
import com.trolltech.qt.gui.QPalette;
import com.trolltech.qt.gui.QPen;
import com.trolltech.qt.gui.QPixmap;
import com.trolltech.qt.gui.QWidget;

public class BalloonWidget extends QWidget {
    
    public enum ColorTheme {
        Neutral,
        Jedi,
        Sith
    }

    private QImage iconImage;
    private QPixmap borderMap;
    private QLabel titleLabel;
    private QLabel msgLabel;
    private final int rectWH = 20;
    private final int arrowW = 30;
    private final int arrowH = 30;
    private final int arrowD = 10; /* arrow shift */
    private final int arrowWD = arrowW+arrowD;
    private final int msgLabelMarginRight = 10;
    private int iconImageW = 48;
    private int iconImageH = 48;
    private boolean arrowTop;
    private boolean arrowLeft;
    private int top = 0;
    private QColor headerColor;
    private QPoint pos;
    private double progress = -1;
    private double progressMax = 0;
    private QImage progressHilt;
    private QImage progressEnd;
    private QImage progress1;
    private int progressPixels;
    private boolean closeOnComplete = true;
    private double pxs; /* pixels per 1 per cent */
    private static QPoint lastToolTipPos;
    private static BalloonWidget lastToolTip;
    private static QWidget lastToolTipWidget;
    private BalloonWidget lastMessage;
    private BalloonWidget lastProgress;
    protected int w = 250;
    protected int h = 70;
    protected int messageTimeout = 3000;
    protected int toolTipTimeout = 3000;
    public Signal0 progressCompleted = new Signal0();
    public Signal0 balloonClicked = new Signal0();
    public Signal0 balloonDoubleClicked = new Signal0();


    public BalloonWidget(QPoint pos) {
        this(pos,ColorTheme.Neutral);
    }

    public BalloonWidget(QPoint pos, ColorTheme colorTheme) {
        setAttribute(Qt.WidgetAttribute.WA_DeleteOnClose);
        setWindowFlags(Qt.WindowType.ToolTip);
        this.pos = pos;

        switch (colorTheme) {
        case Sith:
            headerColor = new QColor(0x982100);
            iconImage = new QImage("images/sith.png");
            progressHilt = new QImage("images/hilt_red.png");
            progressEnd = new QImage("images/end_red.png");
            progress1 = new QImage("images/1_red.png");
            break;

        case Jedi:
            headerColor = new QColor(0x219800);            
            iconImage = new QImage("images/jedi.png");
            progressHilt = new QImage("images/hilt_green.png");
            progressEnd = new QImage("images/end_green.png");
            progress1 = new QImage("images/1_green.png");
            break;

        default: /* Neutral */
            headerColor = new QColor(0x002198);
            iconImage = new QImage("images/neutral.png");
            progressHilt = new QImage("images/hilt_blue.png");
            progressEnd = new QImage("images/end_blue.png");
            progress1 = new QImage("images/1_blue.png");
            break;
        }
        
        if (iconImage == null || iconImage.width() != iconImageW || iconImage.height() != iconImageH) {
            iconImage = new QImage(48,48,QImage.Format.Format_ARGB32);
            iconImage.fill(QColor.transparent.rgba());
        }
        
        if (progressHilt.width() < 1 || progressEnd.width() < 1 || progress1.width() < 1) {
            progressHilt = null;
            progressEnd = null;
            progress1 = null;
        }
        
        progressPixels = w-iconImageW-progressHilt.width()-30;
        pxs = progressPixels/100.0;

        arrowTop = pos.y() - h - arrowH < 0;
        arrowLeft = pos.x() - w + arrowD + rectWH < 0;
        
        drawBalloon();
        
        resize(w+10,h+10+arrowH);
        QPalette pal = palette();
        pal.setColor(QPalette.ColorRole.Window, new QColor(0xffffff));
        setPalette(pal);
        
        titleLabel = new QLabel(this);
        titleLabel.move(10, 5+top);
        titleLabel.setTextFormat(Qt.TextFormat.PlainText);
        QFont f = titleLabel.font();
        f.setBold(true);
        titleLabel.setFont(f);
        pal = titleLabel.palette();
        pal.setColor(QPalette.ColorRole.WindowText, QColor.white);
        titleLabel.setPalette(pal);
        
        msgLabel = new QLabel(this);
        msgLabel.setTextFormat(Qt.TextFormat.PlainText);
        QContentsMargins cm = msgLabel.getContentsMargins();
        cm.right = msgLabelMarginRight;
        cm.bottom = 15;
        cm.top = 0;
        msgLabel.setContentsMargins(cm);
        msgLabel.resize(w-cm.right-iconImage.width()-10, msgLabel.height());
        msgLabel.setWordWrap(true);
        msgLabel.move(iconImage.width()+10, 25+top);
        f = msgLabel.font();
        f.setBold(true);
        titleLabel.setFont(f);
        pal = msgLabel.palette();
        pal.setColor(QPalette.ColorRole.WindowText, new QColor(0x494949));
        msgLabel.setPalette(pal);
    }
    
    protected void paintEvent(QPaintEvent e) {
        QPainter p = new QPainter(this);
        p.drawPixmap(0, 0, borderMap);
        p.fillRect(0, 0, w+1, 20+top, new QBrush(headerColor));
        p.drawImage(0,h/2-iconImage.height()/2+10+top/2, iconImage);
        if (progress > -1 && progressHilt != null) {
            double pxr = 0; /* unprinted floating point data from previous pixel */
            int px = iconImage.width()+10+progressHilt.width();
            p.drawImage(iconImage.width()+10, h-15, progressHilt);
            for (int i = 0; i < (int)progress; i++) {
                int j = 0;
                for (; j < (int)(pxs+pxr); j++, px++) {
                    p.drawImage(px, h-15, progress1);
                }
                pxr += pxs-j;
            }
            p.drawImage(px, h-15, progressEnd);
        }
        p.end();
    }
    
    protected void mouseReleaseEvent(QMouseEvent e) {
        if (e.button() == Qt.MouseButton.LeftButton) {
            balloonClicked.emit();
        }
        close();
    }
    
    protected void mouseDoubleClickEvent(QMouseEvent e) {
        if (e.button() == Qt.MouseButton.LeftButton) {
            balloonDoubleClicked.emit();
        }
    }
    
    protected void closeEvent(QCloseEvent e) {
        lastToolTipWidget = null;
        lastMessage = null;
        lastProgress = null;
        super.closeEvent(e);
    }
    
    private void drawBalloon() {
        if (arrowTop) {
            top = arrowH;
            h += top;
        }
        
        QPainterPath p = new QPainterPath();
        QRectF r = new QRectF(0,top,rectWH,rectWH);

        p.moveTo(0, rectWH+top);
        p.arcTo(r, 180, -90);
        /* arrow top left */
        if (arrowTop && arrowLeft) {
            move(pos.x()-rectWH-arrowD,pos.y());
            p.lineTo(rectWH, top);
            p.lineTo(rectWH+arrowD, top-arrowH);
            p.lineTo(rectWH+arrowWD, top);
        }
        /* arrow top right */
        else
        if (arrowTop && !arrowLeft) {
            move(pos.x()-w+rectWH+arrowD,pos.y());
            p.lineTo(w-rectWH-arrowWD, top);
            p.lineTo(w-rectWH-arrowD, top-arrowH);
        }
        p.lineTo(w-rectWH, top);
        r.moveTo(w-rectWH, top);

        p.arcTo(r, 90, -90);
        p.lineTo(w, h-rectWH);
        r.moveTo(w-rectWH, h-rectWH);

        p.arcTo(r, 0, -90);
        /* arrow bottom right */
        if (!arrowTop && !arrowLeft) {
            move(pos.x()-w+rectWH+arrowD,pos.y()-h-arrowH);
            p.lineTo(w-rectWH, h);
            p.lineTo(w-rectWH-arrowD, h+arrowH);
            p.lineTo(w-rectWH-arrowWD, h);
        }
        /* arrow bottom left */
        else
        if (!arrowTop && arrowLeft) {
            move(pos.x()-rectWH-arrowD,pos.y()-h-arrowH);
            p.lineTo(rectWH+arrowWD, h);
            p.lineTo(rectWH+arrowD, h+arrowH);
        }
        p.lineTo(rectWH, h);
        r.moveTo(0, h-rectWH);
        
        p.arcTo(r, 270, -90);
        p.lineTo(0,rectWH+top);
        
        /* balloon mask */
        QBitmap bitmap = new QBitmap(800,600);
        bitmap.fill(QColor.color0);
        QPainter painter = new QPainter(bitmap);
        painter.setPen(QColor.color1);
        painter.setBrush(QColor.color1);
        painter.drawPath(p);
        painter.end();
        setMask(bitmap);

        /* balloon border */
        borderMap = new QPixmap(800,600);
        painter = new QPainter(borderMap);
        painter.setPen(new QPen(headerColor,1));
        painter.setBrush(QColor.color0);
        painter.drawPath(p);
        painter.end();
    }
    
    public boolean canShowMessage() {
        return lastMessage == null;
    }
    
    public boolean canShowProgress() {
        return lastProgress != null;
    }
    
    private void updateLabels() {
//        titleLabel.updateGeometry();
//        titleLabel.update(titleLabel.contentsRect());
//        msgLabel.updateGeometry();
//        msgLabel.update(msgLabel.contentsRect());
    }
    
    public void showMessage(final String title, final String msg, int msec) {
        titleLabel.setText(title);
        msgLabel.setText(msg);
        updateLabels();
        lastMessage = this;
        show();
        if (msec > 0) {
            QTimer.singleShot(msec, this, "timerTimeout()");
        }
    }
    
    public void showMessage(final String title, final String msg) {
        showMessage(title, msg, messageTimeout);
    }
    
    public void setProgress(final String title, final String msg, double max) {
        setProgress(title, msg, max, true);
    }

    public void setProgress(final String title, final String msg, double max, boolean closeOnComplete) {
        titleLabel.setText(title);
        msgLabel.setText(msg);
        updateLabels();
        this.progressMax = max;
        this.progress = 0;
        this.closeOnComplete = closeOnComplete;
        lastProgress = this;
        show();
    }

    public void showProgress(final String title, final String msg, double progress) {
        titleLabel.setText(title);
        msgLabel.setText(msg);
        updateLabels();
        showProgress(progress);
    }

    public void showProgress(double progress) {
        if (progress >= 0) {
            if (progress < progressMax) {
                this.progress = (100*progress)/progressMax;
                repaint();
            }
            else {
                this.progress = 100;
                repaint();
                this.progress = -1;
                this.progressMax = 0;
                progressCompleted.emit();
                if (closeOnComplete) {
                    close();
                }
            }
        }
    }

    @SuppressWarnings("unused")
    private void timerTimeout() {
        close();
    }

    static public boolean canShowToolTip(final QWidget w, QPoint pos) {
        if (lastToolTipPos == null || !lastToolTipPos.equals(pos)
                && (lastToolTipWidget == null || lastToolTipWidget != w)) {
            return true;
        }
        return false;
    }
    
    public void showToolTip(final QWidget w, final String title, final String toolTip, int msec) {
        if (canShowToolTip(w,pos)) {
            lastToolTipWidget = w;
            if (lastToolTip != null) {
                lastToolTip.dispose();
            }
            lastToolTip = this;
            lastToolTipPos = pos;
            titleLabel.setText(title);
            msgLabel.setText(toolTip);
            updateLabels();
            show();
            if (msec > 0) {
                QTimer.singleShot(msec, this, "timerTimeout()");
            }
        }
    }
    
    public void showToolTip(final QWidget w, final String title, final String toolTip) {
        showToolTip(w, title, toolTip, toolTipTimeout);
    }

    public void showToolTip(final QWidget w, final String toolTip, int msec) {
        showToolTip(w, "ToolTip Message", toolTip, msec);
    }
    
    public void showToolTip(final QWidget w, final String toolTip) {
        showToolTip(w, "ToolTip Message", toolTip, toolTipTimeout);
    }

    public void showToolTip(final QWidget w, int msec) {
        showToolTip(w, "ToolTip Message", w.toolTip(), msec);
    }

    public void showToolTip(final QWidget w) {
        showToolTip(w, "ToolTip Message", w.toolTip(), toolTipTimeout);
    }
}

File: src/gui/BalloonWidget.cpp
#include "gui/BalloonWidget.h"

#include <QBitmap>
#include <QCloseEvent>
#include <QCoreApplication>
#include <QFont>
#include <QMouseEvent>
#include <QPainter>
#include <QPainterPath>
#include <QPaintEvent>
#include <QPalette>
#include <QPen>
#include <QScreen>
#include <QTimer>

// Static members
QPoint BalloonWidget::lastToolTipPos_;
BalloonWidget *BalloonWidget::lastToolTip_ = nullptr;
QWidget *BalloonWidget::lastToolTipWidget_ = nullptr;

BalloonWidget::BalloonWidget(QPoint pos, ColorTheme colorTheme)
    : QWidget(nullptr)
    , pos_(pos)
{
    setAttribute(Qt::WA_DeleteOnClose);
    setWindowFlags(Qt::ToolTip);

    QString imgDir = QCoreApplication::applicationDirPath() + "/images/";

    switch (colorTheme) {
    case Sith:
        headerColor_ = QColor(0x982100);
        iconImage_ = QImage(imgDir + "sith.png");
        progressHilt_ = QImage(imgDir + "hilt_red.png");
        progressEnd_ = QImage(imgDir + "end_red.png");
        progress1_ = QImage(imgDir + "1_red.png");
        break;

    case Jedi:
        headerColor_ = QColor(0x219800);
        iconImage_ = QImage(imgDir + "jedi.png");
        progressHilt_ = QImage(imgDir + "hilt_green.png");
        progressEnd_ = QImage(imgDir + "end_green.png");
        progress1_ = QImage(imgDir + "1_green.png");
        break;

    default: // Neutral
        headerColor_ = QColor(0x002198);
        iconImage_ = QImage(imgDir + "neutral.png");
        progressHilt_ = QImage(imgDir + "hilt_blue.png");
        progressEnd_ = QImage(imgDir + "end_blue.png");
        progress1_ = QImage(imgDir + "1_blue.png");
        break;
    }

    if (iconImage_.isNull() || iconImage_.width() != ICON_IMAGE_W || iconImage_.height() != ICON_IMAGE_H) {
        iconImage_ = QImage(48, 48, QImage::Format_ARGB32);
        iconImage_.fill(Qt::transparent);
    }

    if (progressHilt_.width() < 1 || progressEnd_.width() < 1 || progress1_.width() < 1) {
        progressHilt_ = QImage();
        progressEnd_ = QImage();
        progress1_ = QImage();
    }

    progressPixels_ = w_ - ICON_IMAGE_W - progressHilt_.width() - 30;
    pxs_ = progressPixels_ / 100.0;

    arrowTop_ = pos_.y() - h_ - ARROW_H < 0;
    arrowLeft_ = pos_.x() - w_ + ARROW_D + RECT_WH < 0;

    drawBalloon();

    resize(w_ + 10, h_ + 10 + ARROW_H);
    QPalette pal = palette();
    pal.setColor(QPalette::Window, QColor(0xffffff));
    setPalette(pal);

    // Title label
    titleLabel_ = new QLabel(this);
    titleLabel_->move(10, 5 + top_);
    titleLabel_->setTextFormat(Qt::PlainText);
    QFont f = titleLabel_->font();
    f.setBold(true);
    titleLabel_->setFont(f);
    pal = titleLabel_->palette();
    pal.setColor(QPalette::WindowText, Qt::white);
    titleLabel_->setPalette(pal);

    // Message label
    msgLabel_ = new QLabel(this);
    msgLabel_->setTextFormat(Qt::PlainText);
    msgLabel_->setContentsMargins(0, 0, MSG_LABEL_MARGIN_RIGHT, 15);
    msgLabel_->resize(w_ - MSG_LABEL_MARGIN_RIGHT - iconImage_.width() - 10, msgLabel_->height());
    msgLabel_->setWordWrap(true);
    msgLabel_->move(iconImage_.width() + 10, 25 + top_);
    f = msgLabel_->font();
    f.setBold(true);
    titleLabel_->setFont(f);
    pal = msgLabel_->palette();
    pal.setColor(QPalette::WindowText, QColor(0x494949));
    msgLabel_->setPalette(pal);
}

void BalloonWidget::paintEvent(QPaintEvent *event)
{
    Q_UNUSED(event);
    QPainter p(this);
    p.drawPixmap(0, 0, borderMap_);
    p.fillRect(0, 0, w_ + 1, 20 + top_, QBrush(headerColor_));
    p.drawImage(0, h_ / 2 - iconImage_.height() / 2 + 10 + top_ / 2, iconImage_);
    if (progress_ > -1 && !progressHilt_.isNull()) {
        double pxr = 0; // unprinted floating point data from previous pixel
        int px = iconImage_.width() + 10 + progressHilt_.width();
        p.drawImage(iconImage_.width() + 10, h_ - 15, progressHilt_);
        for (int i = 0; i < static_cast<int>(progress_); i++) {
            int j = 0;
            for (; j < static_cast<int>(pxs_ + pxr); j++, px++) {
                p.drawImage(px, h_ - 15, progress1_);
            }
            pxr += pxs_ - j;
        }
        p.drawImage(px, h_ - 15, progressEnd_);
    }
    p.end();
}

void BalloonWidget::mouseReleaseEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        emit balloonClicked();
    }
    close();
}

void BalloonWidget::mouseDoubleClickEvent(QMouseEvent *event)
{
    if (event->button() == Qt::LeftButton) {
        emit balloonDoubleClicked();
    }
}

void BalloonWidget::closeEvent(QCloseEvent *event)
{
    lastToolTipWidget_ = nullptr;
    lastMessage_ = nullptr;
    lastProgress_ = nullptr;
    QWidget::closeEvent(event);
}

void BalloonWidget::drawBalloon()
{
    if (arrowTop_) {
        top_ = ARROW_H;
        h_ += top_;
    }

    QPainterPath p;
    QRectF r(0, top_, RECT_WH, RECT_WH);

    p.moveTo(0, RECT_WH + top_);
    p.arcTo(r, 180, -90);

    // Arrow top left
    if (arrowTop_ && arrowLeft_) {
        move(pos_.x() - RECT_WH - ARROW_D, pos_.y());
        p.lineTo(RECT_WH, top_);
        p.lineTo(RECT_WH + ARROW_D, top_ - ARROW_H);
        p.lineTo(RECT_WH + ARROW_WD, top_);
    }
    // Arrow top right
    else if (arrowTop_ && !arrowLeft_) {
        move(pos_.x() - w_ + RECT_WH + ARROW_D, pos_.y());
        p.lineTo(w_ - RECT_WH - ARROW_WD, top_);
        p.lineTo(w_ - RECT_WH - ARROW_D, top_ - ARROW_H);
    }

    p.lineTo(w_ - RECT_WH, top_);
    r.moveTo(w_ - RECT_WH, top_);

    p.arcTo(r, 90, -90);
    p.lineTo(w_, h_ - RECT_WH);
    r.moveTo(w_ - RECT_WH, h_ - RECT_WH);

    p.arcTo(r, 0, -90);

    // Arrow bottom right
    if (!arrowTop_ && !arrowLeft_) {
        move(pos_.x() - w_ + RECT_WH + ARROW_D, pos_.y() - h_ - ARROW_H);
        p.lineTo(w_ - RECT_WH, h_);
        p.lineTo(w_ - RECT_WH - ARROW_D, h_ + ARROW_H);
        p.lineTo(w_ - RECT_WH - ARROW_WD, h_);
    }
    // Arrow bottom left
    else if (!arrowTop_ && arrowLeft_) {
        move(pos_.x() - RECT_WH - ARROW_D, pos_.y() - h_ - ARROW_H);
        p.lineTo(RECT_WH + ARROW_WD, h_);
        p.lineTo(RECT_WH + ARROW_D, h_ + ARROW_H);
    }

    p.lineTo(RECT_WH, h_);
    r.moveTo(0, h_ - RECT_WH);

    p.arcTo(r, 270, -90);
    p.lineTo(0, RECT_WH + top_);

    // Balloon mask
    QBitmap bitmap(800, 600);
    bitmap.fill(Qt::color0);
    {
        QPainter painter(&bitmap);
        painter.setPen(Qt::color1);
        painter.setBrush(Qt::color1);
        painter.drawPath(p);
    }
    setMask(bitmap);

    // Balloon border
    borderMap_ = QPixmap(800, 600);
    {
        QPainter painter(&borderMap_);
        painter.setPen(QPen(headerColor_, 1));
        painter.setBrush(Qt::color0);
        painter.drawPath(p);
    }
}

bool BalloonWidget::canShowMessage() const
{
    return lastMessage_ == nullptr;
}

bool BalloonWidget::canShowProgress() const
{
    return lastProgress_ != nullptr;
}

void BalloonWidget::showMessage(const QString &title, const QString &msg, int msec)
{
    titleLabel_->setText(title);
    msgLabel_->setText(msg);
    lastMessage_ = this;
    show();
    if (msec > 0) {
        QTimer::singleShot(msec, this, &BalloonWidget::close);
    }
}

void BalloonWidget::showMessage(const QString &title, const QString &msg)
{
    showMessage(title, msg, messageTimeout_);
}

void BalloonWidget::setProgress(const QString &title, const QString &msg, double max, bool closeOnComplete)
{
    titleLabel_->setText(title);
    msgLabel_->setText(msg);
    progressMax_ = max;
    progress_ = 0;
    closeOnComplete_ = closeOnComplete;
    lastProgress_ = this;
    show();
}

void BalloonWidget::showProgress(const QString &title, const QString &msg, double progress)
{
    titleLabel_->setText(title);
    msgLabel_->setText(msg);
    showProgress(progress);
}

void BalloonWidget::showProgress(double progress)
{
    if (progress >= 0) {
        if (progress < progressMax_) {
            progress_ = (100 * progress) / progressMax_;
            repaint();
        } else {
            progress_ = 100;
            repaint();
            progress_ = -1;
            progressMax_ = 0;
            emit progressCompleted();
            if (closeOnComplete_) {
                close();
            }
        }
    }
}

bool BalloonWidget::canShowToolTip(QWidget *w, QPoint pos)
{
    if (lastToolTipPos_.isNull() || lastToolTipPos_ != pos) {
        return true;
    }
    if (lastToolTipWidget_ == nullptr || lastToolTipWidget_ != w) {
        return true;
    }
    return false;
}

void BalloonWidget::showToolTip(QWidget *w, const QString &title, const QString &toolTip, int msec)
{
    if (canShowToolTip(w, pos_)) {
        lastToolTipWidget_ = w;
        if (lastToolTip_ != nullptr) {
            lastToolTip_->close();
            lastToolTip_ = nullptr;
        }
        lastToolTip_ = this;
        lastToolTipPos_ = pos_;
        titleLabel_->setText(title);
        msgLabel_->setText(toolTip);
        show();
        if (msec > 0) {
            QTimer::singleShot(msec, this, &BalloonWidget::close);
        }
    }
}

void BalloonWidget::showToolTip(QWidget *w, const QString &title, const QString &toolTip)
{
    showToolTip(w, title, toolTip, toolTipTimeout_);
}

void BalloonWidget::showToolTip(QWidget *w, const QString &toolTip, int msec)
{
    showToolTip(w, QStringLiteral("ToolTip Message"), toolTip, msec);
}

void BalloonWidget::showToolTip(QWidget *w, const QString &toolTip)
{
    showToolTip(w, QStringLiteral("ToolTip Message"), toolTip, toolTipTimeout_);
}

void BalloonWidget::showToolTip(QWidget *w, int msec)
{
    showToolTip(w, QStringLiteral("ToolTip Message"), w->toolTip(), msec);
}

void BalloonWidget::showToolTip(QWidget *w)
{
    showToolTip(w, QStringLiteral("ToolTip Message"), w->toolTip(), toolTipTimeout_);
}
